mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+#![feature(proc_macro)]
+extern crate proc_macro;
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use syn::{ Body, VariantData };
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(MagicDom)]
+pub fn magic_dom(input: TokenStream) -> TokenStream {
+    let s = input.to_string();
+    let ast = syn::parse_macro_input(&s).unwrap();
+
+	let gen = match ast.body {
+		Body::Enum(_) => panic!("#[derive(magic-dom)] can only be used with structs"),
+		Body::Struct(ref data) => match_struct(&ast, data),
+	};
+    gen.to_string().parse().unwrap() 
+}
+
+fn match_struct( ast: &syn::DeriveInput, variant: &syn::VariantData ) -> quote::Tokens {
+    let name_spec = &ast.ident;
+    let name_spec_str = name_spec.to_string();
+    let split_name: Vec<&str> = name_spec_str.split('_').collect();
+    let name = quote::Ident::from(split_name[0]);
+	let js_class = quote::Ident::from(format!("{}_class", name));
+	let struct_def_gc = get_struct_def_gc(variant);
+	let (struct_def, slot_num) = get_struct_def(variant);
+    let struct_field_tests = get_addr_test_struct_field(variant);
+    let test_fn_name = quote::Ident::from(format!("test_{}_magic_layout()", name));
+    
+    // TODO: Need to generate the following implementation
+    // const #js_class: jsapi::js::Class = {
+    // }
+    quote! {
+		struct #name {
+			#(#struct_def,)*
+		}
+
+
+		impl RootKind for #name  {
+			#[inline(always)]
+			fn rootKind() -> jsapi::JS::RootKind {
+				jsapi::JS::RootKind::Object
+			}
+		}
+
+		impl GCMethods for #name  {
+			unsafe fn initial() -> #name {
+				#name {
+					#(#struct_def_gc,)*	
+				}
+			}
+
+			unsafe fn post_barrier(v: *mut #name, prev: #name, next: #name) {
+				unimplemented!()
```

Comment:
You'll need to write the `as_jsobject` method yourself.

It won't lead to infinite recursion because it is calling a different `GCMethods::post_barrier` implementation. The one being defined is `GCMethods::<$magic_dom_struct_type>::post_barrier`, which will be forwarding to the `GCMethods::<*mut jsapi::JSObject>::post_barrier` implementation, which ultimately calls `jsapi::JS::HeapObjectPostBarrier`.